Scrum 50 add logout endpoint

SergeyGrinenko opened the pull request:
### Summary (what was done)

- Implemented Logout flow for IAM: DELETE /api/token returns 204 No Content.
- Added refresh token revocation in MySQL via refresh_tokens table (migration included).
- Introduced RefreshTokenHasher (store only token hashes) and DbalRefreshTokenStore (atomic revoke via single UPDATE).
- Added domain error types: AUTH_UNAUTHORIZED, AUTH_TOKEN_INVALID, VALIDATION_ERROR.
- Added ExceptionSubscriber to map domain errors to consistent HTTP + JSON responses.
- Added UserIdAwareInterface to ensure authenticated user exposes getId() (type-safety / PHPStan).
- Added dev CLI commands for manual testing: seed refresh token + run logout use case.

### How it was tested


<img width="1647" height="840" alt="1 1" src="https://github.com/user-attachments/assets/18190c4f-ae65-4714-93f6-92bc97bf6d3c" />
<img width="1655" height="304" alt="Знімок екрана 2025-12-18 о 11 59 39" src="https://github.com/user-attachments/assets/c3e162d3-1372-4381-bf02-3f9afbde7401" />
<img width="1647" height="102" alt="1 2" src="https://github.com/user-attachments/assets/326fe267-11a2-4778-bbee-33a5ef4f0693" />
<img width="884" height="464" alt="1 3" src="https://github.com/user-attachments/assets/ee8e2362-fbfc-4193-a185-9e5124a38e73" />

SergeyGrinenko committed:
chore(dependencies): SCRUM-50 add Symfony SecurityBundle dependencies

Added required Symfony Security stack dependencies (pulled transitively) so authentication-related features can be wired and the app boots consistently.

SergeyGrinenko committed:
chore(security): SCRUM-50 configure firewall and security routes

Enabled the main firewall and register security-related routes so protected endpoints can resolve an authenticated user in runtime.

SergeyGrinenko committed:
feat(shared): SCRUM-50 add UUID binary converter

Provided a single reusable utility to convert UUID strings to BINARY(16) (and back if needed), matching the DB schema and preventing duplicated conversion logic.

SergeyGrinenko committed:
feat(iam): SCRUM-50 add auth domain exceptions

Introduced stable domain-level error types (AUTH_UNAUTHORIZED, AUTH_TOKEN_INVALID, VALIDATION_ERROR) to keep error handling consistent across controllers.

SergeyGrinenko committed:
feat(iam): SCRUM-50 introduce UserIdAwareInterface

Guarantee that an authenticated user object can expose getId() in a type-safe way (fixes PHPStan and avoids runtime ambiguity from UserInterface).

SergeyGrinenko committed:
feat(iam): SCRUM-50 add refresh token hasher

Ensured refresh tokens are never stored in plain text; persist only a fixed-size hash suitable for BINARY(32) in MySQL.

SergeyGrinenko committed:
feat(iam): SCRUM-50 implement logout use case

Implemented the application-level Logout flow (business logic) as a use case that revokes a refresh token, independent of HTTP/UI.

SergeyGrinenko committed:
feat(iam): SCRUM-50 add DBAL refresh token store and migration

Added MySQL persistence for refresh tokens (migration + DBAL store) so logout can revoke tokens in a real database table.

SergeyGrinenko committed:
feat(iam): SCRUM-50 add exception subscriber for domain errors

Centralized mapping of domain exceptions to HTTP status codes and JSON error responses, so controllers stay thin and responses are consistent.

SergeyGrinenko committed:
feat(iam): SCRUM-50 add DELETE /api/token logout endpoint

Logout via REST (DELETE /api/token) with input validation and a clean 204 No Content response on success.

SergeyGrinenko committed:
chore(dev): SCRUM-50 add CLI commands for refresh token testing

Provided developer-only console commands to seed refresh tokens and manually test logout without needing the login endpoint yet.

SergeyGrinenko committed:
test(iam): SCRUM-50 cover logout flow

Added automated tests to verify logout behavior and regression-proof the revoke logic / error scenarios.

SergeyGrinenko committed:
fixup! feat(iam): SCRUM-50 add DBAL refresh token store and migration

Refactor migration SQL to heredoc / formatting cleanup.

SergeyGrinenko committed:
fixup! feat(iam): SCRUM-50 add DBAL refresh token store and migration

use IF EXISTS for refresh_tokens drop

CPAprince commented on src/IAM/Infrastructure/Persistence/MySQL/Logout/DbalRefreshTokenStore.php:
```diff
@@ -0,0 +1,51 @@
+UPDATE refresh_tokens
+SET revoked_at = :now
+WHERE token_hash = :hash
+  AND user_id = :userId
```
Why something else in WHERE statement exept token_hash?
Is there many rows where `token_hash` the same? 
`UNIQUE INDEX uniq_refresh_token_hash (token_hash),` it's unique field


Seems like 
`UPDATE refresh_tokens
SET revoked_at = :now
WHERE token_hash = :hash`
do the same job whithout unnecessary additional loading

SergeyGrinenko replied:
token_hash is really unique, so yes - WHERE token_hash = :hash would find the row.
We still keep AND user_id = :userId like a "safety check" logout runs for authenticated user, and we revoke only token related to that user. This prevents revoking someone else’s token if a foreign refresh token is leaked and sent with a different user session.

Yes, probably it's seems like overcautious, but, it’s still one UPDATE (no extra query), just an extra guard condition.

CPAprince reviewed (changes requested):
SQL Where statement could be shorter

SergeyGrinenko committed:
fixup! feat(iam): SCRUM-50 add exception subscriber for domain errors

Limited exception subscriber to /api requests to avoid converting non-API routes errors into JSON

SergeyGrinenko committed:
fixup! feat(iam): SCRUM-50 add DELETE /api/token logout endpoint

- Moved logout route to DELETE /api/tokens
- Enforced application/json with 415 Unsupported Media Type

ihpr commented on src/IAM/Application/Logout/LogoutService.php:
```diff
@@ -0,0 +1,17 @@
+    {
+    }
+
+    public function logout(string $userId, string $refreshToken): void
```
@SergeyGrinenko Check the CQRS pattern in the documentation.

CC: @adnlv Add query example to the module documentation

ihpr commented on src/IAM/Infrastructure/Http/ExceptionSubscriber.php:
```diff
@@ -0,0 +1,54 @@
+            return;
+        }
+
+        $e = $event->getThrowable();
```
@SergeyGrinenko Avoid one-letter names.

ihpr commented on src/IAM/UI/REST/Controller/TokenController.php:
```diff
@@ -0,0 +1,60 @@
+        }
+
+        if (!$user instanceof UserIdAwareInterface) {
+            throw new \RuntimeException('Authenticated user must implement UserIdAwareInterface (getId()).');
```
@SergeyGrinenko Simplify FQN.

ihpr commented on src/IAM/UI/REST/Controller/TokenController.php:
```diff
@@ -0,0 +1,60 @@
+        }
+
+        try {
+            $payload = json_decode($request->getContent(), true, 512, JSON_THROW_ON_ERROR);
```
@SergeyGrinenko Try to minimize amount of arguments by using named arguments.

ihpr commented on src/IAM/UI/REST/Controller/TokenController.php:
```diff
@@ -0,0 +1,60 @@
+        try {
+            $payload = json_decode($request->getContent(), true, 512, JSON_THROW_ON_ERROR);
+        } catch (\JsonException) {
+            throw new ValidationErrorException('Invalid JSON.');
```
@SergeyGrinenko Invalid JSON is a format issue, which results in a 400 status code.

ihpr commented on src/IAM/UI/REST/Controller/TokenController.php:
```diff
@@ -0,0 +1,60 @@
+
+        $this->logoutService->logout($user->getId(), $refreshToken);
+
+        return new Response(null, 204);
```
@SergeyGrinenko 204 is a constant in the Response class.

ihpr commented on src/IAM/UI/REST/Controller/TokenController.php:
```diff
@@ -0,0 +1,60 @@
+    #[Route('/api/tokens', name: 'api_tokens_logout', methods: ['DELETE'])]
+    public function logout(Request $request): Response
+    {
+        $user = $this->security->getUser();
```
@SergeyGrinenko Validation should be split from the controller logic.

ihpr commented on src/IAM/UI/REST/Controller/TokenController.php:
```diff
@@ -0,0 +1,60 @@
+            throw new AuthUnauthorizedException();
+        }
+
+        if (!$user instanceof UserIdAwareInterface) {
```
@SergeyGrinenko Use CQRS pattern. Command adn command handler must be added

ihpr commented on src/Shared/Infrastructure/Persistence/Doctrine/UuidBinaryConverter.php:
```diff
@@ -0,0 +1,24 @@
+{
+    public static function toBytes(string $uuid): string
+    {
+        $hex = strtolower(str_replace('-', '', $uuid));
```
@SergeyGrinenko Nice to have: https://youtu.be/zXd8iipgcAI?si=fNBsD5zYBSbpyGZN

ihpr commented on src/Shared/Infrastructure/Persistence/UuidBinaryConverter.php:
```diff
@@ -0,0 +1,24 @@
+        }
+
+        $bytes = hex2bin($hex);
+        if (false === $bytes || 16 !== strlen($bytes)) {
```
@SergeyGrinenko Avoid magic numbers like 16

ihpr commented on tests/IAM/Application/Logout/LogoutServiceTest.php:
```diff
@@ -0,0 +1,27 @@
+
+final class LogoutServiceTest extends TestCase
+{
+    public function testLogoutCallsStoreRevoke(): void
```
@SergeyGrinenko Check Artem's PR for a better test format. Use attributes and groups.

ihpr commented on src/IAM/Domain/Auth/Exception/ValidationErrorException.php:
```diff
@@ -0,0 +1,15 @@
+{
+    public const ERROR_CODE = 'VALIDATION_ERROR';
+
+    public function __construct(string $message = 'Validation error.')
```
@SergeyGrinenko Remove default arguemnt. Force message set.

ihpr commented on src/IAM/Infrastructure/Persistence/Auth/RefreshTokenHasher.php:
```diff
@@ -0,0 +1,17 @@
+
+namespace Twitter\IAM\Infrastructure\Persistence\Auth;
+
+final class RefreshTokenHasher
```
@SergeyGrinenko Maybe the whole class can be readonly?

ihpr commented on src/IAM/Infrastructure/Persistence/MySQL/Logout/DbalRefreshTokenStore.php:
```diff
@@ -0,0 +1,51 @@
+
+        $affected = $this->connection->executeStatement(
+            <<<SQL
+UPDATE refresh_tokens
```
@SergeyGrinenko Indents.

ihpr commented on src/IAM/Infrastructure/Persistence/MySQL/Migrations/Version20251217080023.php:
```diff
@@ -0,0 +1,42 @@
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
```
@SergeyGrinenko Delete auto-generated.

ihpr commented on src/IAM/UI/CLI/Dev/LogoutTestCommand.php:
```diff
@@ -0,0 +1,46 @@
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $userId = (string) $input->getArgument('userId');
```
@SergeyGrinenko Add some basic validation.

ihpr commented on src/IAM/UI/CLI/Dev/SeedRefreshTokenTestCommand.php:
```diff
@@ -0,0 +1,62 @@
+        $expiresAt = (new \DateTimeImmutable())->modify('+1 day');
+        $now = new \DateTimeImmutable();
+
+        $this->connection->insert('refresh_tokens', [
```
@SergeyGrinenko How can I reuse that insert?

ihpr commented on src/IAM/UI/CLI/Dev/SeedRefreshTokenCommand.php:
```diff
@@ -0,0 +1,62 @@
+use Twitter\Shared\Infrastructure\Persistence\UuidBinaryConverter;
+
+#[AsCommand(name: 'app:seed-refresh', description: 'Seed one refresh token row for logout testing')]
+final class SeedRefreshTokenCommand extends Command
```
@SergeyGrinenko Mark this class as to be removed

ihpr replied:
@CPAprince Think how you would track tech debt.

ihpr commented on tests/IAM/Infrastructure/Persistence/MySQL/Logout/DbalRefreshTokenStoreTest.php:
```diff
@@ -0,0 +1,98 @@
+        $expectedHash = hash_hmac('sha256', $refreshToken, $secret, true); // 32 bytes
+        $expectedUserBytes = UuidBinaryConverter::toBytes($userId); // 16 bytes
+
+        $connection->expects(self::once())
```
@SergeyGrinenko Remove.

adnlv commented on src/IAM/UI/REST/Controller/TokenController.php:
```diff
@@ -0,0 +1,60 @@
+use Twitter\IAM\Domain\Auth\Exception\ValidationErrorException;
+use Twitter\IAM\Infrastructure\Security\UserIdAwareInterface;
+
+final class TokenController
```
We use [invokable controllers](https://symfony.com/doc/current/controller/service.html#invokable-controllers). See if you could rename the controller to reflect its purpose accurately. For example:

```suggestion
final class RevokeRefreshTokenController
```

adnlv replied:
All properties are read-only, so maybe make the controller read-only as well?

```suggestion
final readonly class RevokeRefreshTokenController
```

adnlv commented on config/packages/doctrine_migrations.yaml:
See https://github.com/CPAprince/Twitter/pull/20#discussion_r2645742209

adnlv commented on config/packages/security.yaml:
Please remove all redundant comments from this file and any others.

adnlv commented on src/IAM/Application/Logout/RefreshTokenStoreInterface.php:
```diff
@@ -0,0 +1,10 @@
+
+namespace Twitter\IAM\Application\Logout;
+
+interface RefreshTokenStoreInterface
```
Is it a repository? If so, please rename it respectfully. For example:

```suggestion
interface RefreshTokenRepository
```

No need for the `Interface` prefix.

adnlv commented on src/IAM/Infrastructure/Persistence/MySQL/Logout/DbalRefreshTokenStore.php:
```diff
@@ -0,0 +1,51 @@
+use Twitter\IAM\Infrastructure\Persistence\Auth\RefreshTokenHasher;
+use Twitter\Shared\Infrastructure\Persistence\UuidBinaryConverter;
+
+final class DbalRefreshTokenStore implements RefreshTokenStoreInterface
```
As specified in the namespace, this implementation is for MySQL. Therefore, specify MySQL in the class name. For example:

```suggestion
final class MySQLRefreshTokenStore implements RefreshTokenStoreInterface
```

SergeyGrinenko committed:
refactor(iam): SCRUM-50 switch logout to command handler

Replaced LogoutService with CQRS-style LogoutCommand + LogoutHandler.
Renamed refresh token store contract to repository and update tests accordingly.

SergeyGrinenko committed:
feat(iam): SCRUM-50 Updated DELETE /api/tokens logout endpoint

Updated logout endpoint that revokes refresh token for authenticated user.
Introduced request DTO/mapper/validator and return 204 on success.

SergeyGrinenko committed:
feat(iam): SCRUM-50 handle domain errors in REST API

Added API exception subscriber for /api routes with consistent JSON error format.
Map domain exceptions to correct HTTP status codes and log safely.

SergeyGrinenko committed:
refactor(iam): SCRUM-50 renamed auth exceptions

Renamed auth exceptions to remove redundant prefixes and make naming consistent.

SergeyGrinenko committed:
refactor(iam): SCRUM-50 moved refresh token infrastructure components

Moved hasher and UUID converter to proper namespaces and update usages.
Rename DBAL refresh token store to MySQLRefreshToken.

SergeyGrinenko committed:
feat(iam): SCRUM-50 added cleanup for revoked expired refresh tokens

Added maintenance command to delete revoked refresh tokens after expiration.
Covered cleanup flow with unit tests.

SergeyGrinenko committed:
chore(dev): SCRUM-50 added CLI helpers for logout testing

Add dev-only commands to seed refresh_tokens and test logout flow manually.
Mark commands as deprecated for later removal.

SergeyGrinenko committed:
fix(iam): SCRUM-50 updated refresh_tokens migration

Cleanup migration for refresh_tokens table and align Doctrine migrations configuration.

SergeyGrinenko committed:
test(iam): SCRUM-50 cover refresh token revoke repository

Updated unit tests for MySQL refresh token revocation and remove obsolete store tests.

SergeyGrinenko committed:
fix(config): SCRUM-50 Removed unnecessary comments

adnlv commented on src/IAM/Domain/Auth/Exception/BadRequestException.php:
```diff
@@ -0,0 +1,15 @@
+
+namespace Twitter\IAM\Domain\Auth\Exception;
+
+final class BadRequestException extends \RuntimeException
```
```suggestion
final class BadRequestException extends RuntimeException
```

Simplify FQNs. For reference, see [afe10ce](https://github.com/CPAprince/Twitter/pull/20/changes/afe10ce7843d0cf1e03526e247450f10540d6159#diff-8ec5bae185dc9af5b5bfdcd5f1dcbee62f18525ee488996d52b22729f4076ef0)

adnlv commented on src/IAM/Infrastructure/Persistence/MySQL/Logout/MySQLRefreshToken.php:
```diff
@@ -0,0 +1,64 @@
+
+declare(strict_types=1);
+
+namespace Twitter\IAM\Infrastructure\Persistence\MySQL\Logout;
```
```suggestion
namespace Twitter\IAM\Infrastructure\Persistence\MySQL\Repository
```

See the [documentation](https://github.com/CPAprince/Twitter/blob/develop/docs/modular-monolith.md#project-structure)

adnlv commented on src/IAM/UI/CLI/Dev/LogoutTestCommand.php:
```diff
@@ -0,0 +1,56 @@
+use Twitter\IAM\Domain\Auth\Exception\TokenInvalidException;
+
+/**
+ * @deprecated Dev-only command. TODO: remove after logout flow is covered by proper fixtures/tests.
```
Why would you want to keep deprecated code in the PR? Add proper unit tests, do not spread unused and unmaintainable code all around the project

ihpr replied:
@adnlv @SergeyGrinenko It's fine to leave a small tech debt in the code. You only need a correct approach to mark it, like creating a tech debt Jira ticket for removing this one and attaching the ticket number and link to the code with a TODO mark. That's ok.

I also see no sense in using the `@deprecetd` annotation (FYI, PHP has an attribute for that) for the newly created class. It's confusing.

adnlv commented on src/IAM/UI/REST/Request/LogoutRequest.php:
```diff
@@ -0,0 +1,13 @@
+final class LogoutRequest
+{
+    public function __construct(
+        public readonly string $refreshToken,
```
Encapsulate properties inside the request and add getters to access them.

```suggestion
        private readonly string $refreshToken,
```

adnlv commented on src/IAM/UI/REST/Request/LogoutRequestMapper.php:
@SergeyGrinenko I think you overdid it. The direction is right, but you could just create a [request class](https://github.com/CPAprince/Twitter/pull/20/changes#diff-760eedfcaeae807b44731397f9ea442d6b364553916a27a9fa9430da06aa7f8aR9) and inject it in the [controller](https://github.com/CPAprince/Twitter/pull/20/changes#diff-51b2b41659c6bd446e75159c200ce63443d3add3fad72274b76bf25173f21a24R28). No need to provide a custom mapper. It is unnecessary complexity.

adnlv commented on src/IAM/UI/REST/Validation/LogoutRequestValidator.php:
1. Use [beberlei/assert](https://github.com/beberlei/assert.git) or [Symfony's validator](https://symfony.com/doc/current/validation.html).
2. Do that in the request class.

adnlv reviewed (changes requested):
Good. Just a few adjustments are required.

SergeyGrinenko committed:
style(php-cs-fixer) SCRUM-50 Simplify FQCNs with imported classes

Updated all touched files by cs-fixer setting

SergeyGrinenko committed:
refactor(logout): SCRUM-50 align folders and namespaces by docs/modular-monolith.md

SergeyGrinenko committed:
chore(gitignore): SCRUM-50 ignore local CLI Dev commands

SergeyGrinenko committed:
refactor(iam): SCRUM-50 encapsulated Logout refresh token

SergeyGrinenko committed:
refactor(iam): SCRUM-50 simplify logout request handling

Removed custom mapper/validator and validate refreshToken in LogoutRequest. Decode JSON in controller and keep controller logic minimal.

SergeyGrinenko committed:
chore(dependencies): SCRUM-50 update Symfony dependencies

SergeyGrinenko committed:
fix(config): SCRUM-50 update RefreshTokenRepository alias

SergeyGrinenko committed:
chore(gitignore): SCRUM-50 remove local CLI Dev commands from repository

ihpr reviewed (commented):
Commented out some reviews. Overall is ok. @SergeyGrinenko Please if you need oen more review round re-request the review. It's a normal continuation after the PR was blocked with `Chanegs requested` status.

adnlv commented on config/services.yaml:
```diff
@@ -1,5 +1,3 @@
-# yaml-language-server: $schema=../vendor/symfony/dependency-injection/Loader/schema/services.schema.json
```
Red Hat YAML Language Server uses specific comments to associate the current YAML file with a specific [JSON schema](https://github.com/redhat-developer/yaml-language-server?tab=readme-ov-file#using-inlined-schema). 

By removing this comment, the IDE will no longer flag mistakes like incorrect data types (e.g., using a string where a number is required) or missing mandatory fields, you will lose context-aware suggestions (via Ctrl+Space) for keys and valid values specific to that file type, hovering over keys will no longer display descriptions or usage instructions that the schema would have provided.

Therefore, please keep the comment.

adnlv commented on src/IAM/Application/Logout/LogoutHandler.php:
```diff
@@ -0,0 +1,18 @@
+
+final class LogoutHandler
+{
+    public function __construct(private readonly RefreshTokenRepository $store) {}
```
```suggestion
    public function __construct(private readonly RefreshTokenRepository $repository) {}
```

@SergeyGrinenko, since you renamed it, it's no longer a "store," is it?

adnlv commented on src/IAM/Infrastructure/Persistence/MySQL/Repository/MySQLRefreshToken.php:
```diff
@@ -0,0 +1,64 @@
+use Twitter\IAM\Infrastructure\Auth\RefreshTokenHasher;
+use Twitter\Shared\Infrastructure\Persistence\Doctrine\UuidBinaryConverter;
+
+final class MySQLRefreshToken implements RefreshTokenRepository
```
```suggestion
final class MySQLRefreshTokenRepository implements RefreshTokenRepository
```

@SergeyGrinenko, be consistent with naming.

adnlv commented on src/IAM/UI/REST/Request/LogoutRequest.php:
```diff
@@ -0,0 +1,23 @@
+        }
+    }
+
+    public function getRefreshToken(): string
```
```suggestion
    public function refreshToken(): string
```

adnlv commented on src/IAM/UI/REST/Request/LogoutRequest.php:
```diff
@@ -0,0 +1,23 @@
+    public function __construct(
+        private string $refreshToken,
+    ) {
+        if ('' === trim($this->refreshToken)) {
```
@SergeyGrinenko, why didn't you want to use [beberlei/assert](https://github.com/beberlei/assert?tab=readme-ov-file#assertthat-chaining)?

This check can be that simple:

```suggestion
        Assert::that($value)->notBlank()
```

adnlv commented on composer.json:
```diff
@@ -81,6 +82,7 @@
     },
     "require-dev": {
         "doctrine/doctrine-fixtures-bundle": "^4.3.1",
+        "friendsofphp/php-cs-fixer": "*",
```
We already have `php-cs-fixer/shim`. Why is there a need for friends of `friendsofphp/php-cs-fixer` as well? Check the [installation guide](https://cs.symfony.com/#installation).

adnlv commented on src/IAM/UI/REST/Controller/RevokeRefreshTokenController.php:
```diff
@@ -0,0 +1,60 @@
+    {
+        $payload = $this->decodeJson($request);
+
+        $logoutRequest = new LogoutRequest(
```
There is no need to parse the body or create a request instance manually. Please pay attention to how the request injection may be implemented: [#20](https://github.com/CPAprince/Twitter/pull/20/changes#diff-51b2b41659c6bd446e75159c200ce63443d3add3fad72274b76bf25173f21a24R27) and [#21](https://github.com/CPAprince/Twitter/pull/21/changes#diff-3c26eec3be738f7effac97cd9144c567552472b2b0640cbc0b4538820356e101R19)

adnlv reviewed (approved):
I won't block the PR, but please pay attention to the suggested changes.

SergeyGrinenko committed:
Merge branch 'develop' into SCRUM-50-Add-Logout-Endpoint

# Conflicts:
#	config/routes.yaml
#	config/services.yaml

CPAprince commented on src/IAM/Infrastructure/Persistence/MySQL/Repository/MySQLRefreshTokenRepository.php:
```diff
@@ -0,0 +1,64 @@
+                SET revoked_at = :now
+                WHERE token_hash = :hash
+                  AND user_id = :userId
+                  AND revoked_at IS NULL
```
```
                  AND user_id = :userId
                  AND revoked_at IS NULL
                  AND expires_at > :now
```
Why do we need this unnecessary additional conditions in Where statement if token_hash is uniq and the only needed key

SergeyGrinenko replied:
token_hash is UNIQUE, so WHERE token_hash = :hash is sufficient to locate the row.

We keep the additional predicates intentionally as guard rails:

user_id = :userId — binds the revoke to the authenticated user context (defense-in-depth).

revoked_at IS NULL — makes the operation idempotent and avoids an unnecessary update on repeated logout.

expires_at > :now — keeps semantics strict: we revoke only active tokens; expired tokens are handled by expiration/cleanup logic.

This does not add noticeable overhead: the row is found via the UNIQUE index on token_hash, and the extra conditions are evaluated on that single row. 

If we prefers a minimal query, I’m fine simplifying

SergeyGrinenko committed:
fix: SCRUM-50 minor fixes

- Minor fixes and cleanup
- Adjust tests and small naming tweaks
- Update composer.lock after changes

CPAprince commented on src/IAM/Infrastructure/Persistence/MySQL/Repository/MySQLRefreshToken.php:
```diff
@@ -0,0 +1,64 @@
+    {
+        return $this->connection->executeStatement(
+            <<<SQL
+                DELETE FROM refresh_tokens
```
Will the table keep expired tokens that wasn't revoked?

SergeyGrinenko replied:
no, but we should setup some cron job for deleteRevokedExpired
then It's will delete expired tokens. For now it's works only manually

CPAprince reviewed (approved):
OK
